packages/x-date-pickers/src/DateField/tests/editing.DateField.test.tsx

```diff
@@ -512,134 +421,150 @@ describe('<DateField /> - Editing', () => {
+          keyStrokes: [
+            // Current query: "J" => 3 matches
+            { value: 'j', expected: '01' },
+            // Current query: "JU" => 2 matches
+            { value: 'u', expected: '06' },
+            // Current query: "JUL" => 1 match
+            { value: 'l', expected: '07' },
+            // Current query: "JULO" => 0 match => fallback set the query to "O"
+            { value: 'o', expected: '10' },
+          ],
+        });
       });
-    });
 
-    it('should not edit when props.readOnly = true and no value is provided', () => {
-      testKeyPress({
-        format: adapterToUse.formats.month,
-        readOnly: true,
-        key: '1',
-        expectedValue: 'MMMM',
+      it('should not edit when props.readOnly = true and no value is provided (letter)', () => {
+        testFieldKeyPress({
+          format: adapter.formats.month,
+          readOnly: true,
+          key: '1',
+          expectedValue: 'MMMM',
+        });
       });
-    });
 
-    it('should not edit value when props.readOnly = true and a value is provided', () => {
-      testKeyPress({
-        format: adapterToUse.formats.month,
-        defaultValue: adapterToUse.date(),
-        readOnly: true,
-        key: 'd',
-        expectedValue: 'June',
+      it('should not edit value when props.readOnly = true and a value is provided (letter)', () => {
+        testFieldKeyPress({
+          format: adapter.formats.month,
+          defaultValue: adapter.date(),
+          readOnly: true,
+          key: 'd',
+          expectedValue: 'June',
+        });
       });
-    });
-  });
+    },
+  );
 
-  describe('Full editing scenarios', () => {
+  describeAdapters('Full editing scenarios', DateField, ({ adapterName, render, clickOnInput }) => {
     it('should move to the last day of the month when the current day exceeds it', () => {
       const onChange = spy();
 
       render(<DateField onChange={onChange} />);
       const input = screen.getByRole('textbox');
       clickOnInput(input, 1);
 
-      fireEvent.change(input, { target: { value: '2 / DD / YYYY' } }); // Press "2"
-      expectInputValue(input, '02 / DD / YYYY');
+      fireEvent.change(input, { target: { value: '1 / DD / YYYY' } }); // Press "1"
+      expectInputValue(input, adapterName === 'luxon' ? '1 / DD / YYYY' : '01 / DD / YYYY');
+
+      fireEvent.change(input, { target: { value: '1 / DD / YYYY' } }); // Press "1"
+      expectInputValue(input, '11 / DD / YYYY');
+
+      fireEvent.change(input, { target: { value: '11 / 3 / YYYY' } }); // Press "3"
+      expectInputValue(input, adapterName === 'luxon' ? '11 / 3 / YYYY' : '11 / 03 / YYYY');
+
+      fireEvent.change(input, { target: { value: '11 / 31 / YYYY' } }); // Press "1"
+      expectInputValue(input, '11 / 31 / YYYY');
+
+      // TODO: Fix this behavior on day.js (`clampDaySection` generates an invalid date for the start of the month).
```

Comment:
`dayjs('01-01-2`, `DD-MM-YYYY`)` returns `Invalid Date`

Note that `dayjs(new Date('0002-01-01').format('YYYY')` returns `2` not `0002`, the error above is not caused by an error in trailing zero.

The best way I see to solve this would be to extract the logic to go from the section string value to the date value (which is in the `setSectionValueOnDate` of `useFieldCharacterEditing` and use `utils.setYear` and `utils.setMonth`

Something like:


```tsx
const startOfMonth = utils.startOfMonth(utils.setMonth(utils.setYear(utils.date()!, getYear(sections)), getMonth(sections)))
```

We would need to make sure it work correctly with 2-digit years and string month.

But the bug is only truly problematic for years before 1000 (`dayjs` parses well `01-01-2022`).
For years after 1000, the only difference with the other adapters is that the day moves when the year is completely filled and not on the first digit.
This is even slightly better since we don't risk to move to the 28th of February when in fact the full year is a leap year.

A cleaner solution would probably be to do the fix described above AND only move to the next day of the month when we want to go to the next section (ie when the year is completely filled).
But if a user types `215` to go to the year `215` then it would never apply the last day of the month since this year could very well be incomplete.

That's clearly not a super important behavior.
Not sure it's worth investing to much time on it, the worst case scenario is if someone writes `31-02-951` it remains `31-02-951` instead of going to `28-02-951` automatically.